### PR TITLE
Shader uniform mvp

### PR DESCRIPTION
**Pull request by CedricGuillemet**

### CODE REVIEW

File: Library/Source/NativeEngine.cpp [hunk 1]
```diff
@@ -753,81 +753,104 @@ namespace babylon
...
+    {
+        const auto uniformData = info[0].As<Napi::External<UniformInfo>>().Data();
+        const auto array = info[1].As<arrayType>();
+
+        size_t elementLength = array.ElementLength();
+
+        m_scratch.clear();
+        for (size_t index = 0; index < elementLength; index += size)
+        {
+            const float values[] = { static_cast<float>(array[index])
```
Comment (bghgary) [inline Library/Source/NativeEngine.cpp]: nit: I don't see this brace style anywhere. Might be good to do it like this:

```C++
const float values[] = {
    static_cast<float>(array[index]),
    (size > 1) ? static_cast<float>(array[index + 1]) : 0.f,
    (size > 2) ? static_cast<float>(array[index + 2]) : 0.f,
    (size > 3) ? static_cast<float>(array[index + 3]) : 0.f
};
```

Comment (bghgary) [reply Library/Source/NativeEngine.cpp]: ditto on the next function

File: Library/Source/NativeEngine.cpp [hunk 2]
```diff
@@ -753,81 +753,104 @@ namespace babylon
...
+    template<int size> void NativeEngine::SetMatrixN(const Napi::CallbackInfo& info)
     {
         const auto uniformData = info[0].As<Napi::External<UniformInfo>>().Data();
         const auto matrix = info[1].As<Napi::Float32Array>();
 
         const size_t elementLength = matrix.ElementLength();
-        assert(elementLength == 16);
+        assert(elementLength == size * size);
+
+        std::array<float, 16> matrixValues{ };
```
Comment (bghgary) [inline Library/Source/NativeEngine.cpp]: nit:

```suggestion
        std::array<float, 16> matrixValues{};
```

File: Library/Source/NativeEngine.cpp [hunk 3]
```diff
@@ -753,81 +753,104 @@ namespace babylon
...
+
+        size_t index = 0;
+        for (int line = 0; line < size; line++)
+        {
+            for (int col = 0; col < size; col++)
+            {
+                matrixValues[line * 4 + col] = matrix[index++];
+            }
+        }
```
Comment (bghgary) [inline Library/Source/NativeEngine.cpp]: Is there a performance issue here when matrix is 4x4? We used to directly use the data, but now we copy the data with a double for loop? Maybe we should specialize the 4x4 case?

**Comment by CedricGuillemet** [comment]
Added the fix for this issue https://github.com/BabylonJS/BabylonNative/issues/85

File: Library/Source/NativeEngine.cpp [hunk 4]
```diff
@@ -770,10 +770,12 @@ namespace babylon
         m_scratch.clear();
         for (size_t index = 0; index < elementLength; index += size)
         {
-            const float values[] = { static_cast<float>(array[index])
+            const float values[] = {
```
Comment (bghgary) [inline Library/Source/NativeEngine.cpp]: nit: brace on newline (sorry, the code I suggest was also wrong style)

File: Library/Source/NativeEngine.cpp [hunk 5]
```diff
@@ -783,10 +785,12 @@ namespace babylon
     template<int size> void NativeEngine::SetFloatN(const Napi::CallbackInfo& info)
     {
         const auto uniformData = info[0].As<Napi::External<UniformInfo>>().Data();
-        const float values[] = { info[1].As<Napi::Number>().FloatValue(),
-                                (size > 1) ? info[2].As<Napi::Number>().FloatValue() : 0.f,
-                                (size > 2) ? info[3].As<Napi::Number>().FloatValue() : 0.f,
-                                (size > 3) ? info[4].As<Napi::Number>().FloatValue() : 0.f };
+        const float values[] = {
```
Comment (bghgary) [inline Library/Source/NativeEngine.cpp]: ditto

**Pull request merged**